runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ConfigBuilder.java

```diff
@@ -131,6 +117,30 @@ public Config build() {
     }
   }
 
+  @VisibleForTesting
+  static Map<String, String> createBundleConfig(
+      SamzaPipelineOptions options, Map<String, String> config) {
+    ImmutableMap.Builder<String, String> builder = ImmutableMap.builder();
+    builder.put(MAX_CONCURRENCY, String.valueOf(options.getMaxBundleSize()));
+
+    if (options.getMaxBundleSize() > 1) {
+      final int threadPoolSize = ConfigUtils.asJobConfig(config).getThreadPoolSize();
+      LOG.info("Remove threadPoolSize configs when maxBundleSize > 1");
+      builder.put(JOB_CONTAINER_THREAD_POOL_SIZE, "0");
+      builder.put(JOB_AUTOSIZING_CONTAINER_THREAD_POOL_SIZE, "0");
+
+      if (threadPoolSize > 1 && options.getNumThreadsForProcessElement() <= 1) {
+        // In case the user sets the thread pool through samza config instead options,
+        // set the bundle thread pool size based on container thread pool config
+        // this allows Samza auto-sizing to tune the threads
+        LOG.info("Convert threadPoolSize {} to numThreadsForProcessElement", threadPoolSize);
+        // NumThreadsForProcessElement in option is the source of truth
+        options.setNumThreadsForProcessElement(threadPoolSize);
```

Comment:
If the user sets the pipeline options in code, it should take higher precedence. Otherwise it will be quite confusing. Autosizing will work if the user doesn't set pipeline options.

We set threadPoolSize to 0 regardless.

Reply:
Isn't this opposite to the existing pattern we have for autosizing? i.e., users can configure resourcing but if they enable autosizing, we treat autosizing overrides as the final values.

This would be inconsistent with the existing behavior for every other auto-sizer tunable configurations.

Reply:
Yes, autosizing will override the same user config. But this is a pipeline option in code. It'll be very confusing if we set some parameters in code while it won't be taken into effect in the runtime. To me, this is definitely higher in precedence. The other problem is that we need a way for user to override without turning off autosizing. Our autosizing hasn't been tuned for thread pool size with bundle > 1. So we need to have this capability around.

Reply:
I feel the inconsistency in behavior is as well confusing to the users. If we want to establish user code as the highest precedence, maybe we should not tie autosizing `job.autosizing.container.thread.pool.size` into this.

Consider this example
1. user configures `job.container.thread.pool.size`, it is reflected in the application behavior
2. autosizing is enabled and user configuration no longer is reflected in the application behavior
3. User sets it in code and the change is reflected in application behavior

user configuration << framework configuration override << user code 

Also consider autosizing devs/support team not realizing knobs x, y and z follows auto-sizer values as higher precedence on their end vs a, b follows user values as higher precedence.


>The other problem is that we need a way for user to override without turning off autosizing
I think the recommended way for now is to use auto-sizer tool to overrides specific knobs.

Reply:
To clarify, ` job.container.thread.pool.size` won't work if auto-sizing is enabled.
auto sizing today allows user overrides size configs but still leave auto sizing enabled (asc can still resize the job)  by:
1. users can configure `job.autosizing.container.thread.pool.size`. It will be honored when the job starts
2. when auto-sizing resizes the job, it will submit the job to jetty proxy with a new value of 
    `job.autosizing.container.thread.pool.size` included in the submission config. and write it to the coordinator 
   stream
3. If user redeploys the job again, the size in 1 is still honored. If the user removes the config in 1 and redeploy, 
   the  size in 2 will be honored (reading from coordinator stream)

for beam job options like ThreadsForProcessElement, asc is not able to achieve 2 so I agree that in ConfigBuilder we should honor the `job.autosizing.container.thread.pool.size` over options specified in code to be able to resize the job. Afterall, if `job.autosizing.container.thread.pool.size` is present, if means either asc resizes the job or the job owner wants to override the size but still want to leave auto-sizing on

Reply:
> I think the recommended way for now is to use auto-sizer tool to overrides specific knobs.
asc-tool does the same thing as step 2 above. ConfigBuilder still needs to address the problem job config vs pipeline option precedence here

Reply:
Synced up offline with @mynameborat  and @jia-gao . We will go with this option precedence approach first to unblock jobs onboarding and then follow up with samza autosizing to figure out precedence.

Reply:
We sync'd up offline on this. The viewpoint to have autosizing enabled plus allow user overrides for knobs that autosizing hasn't caught up on is very valid scenario.

We don't have such compartmentalization today as there are relationships between policies dealing with different resources and hence disabling autosizing all out is the current recommendation to override. That said, unblocking this PR for site-up and we can follow up on this